app/src/main/java/com/seafile/seadroid2/data/StorageManager.java

```diff
@@ -0,0 +1,517 @@
+
+        File[] dirs = getSystemMediaDirs();
+        File[] cacheDirs = getSystemCacheDirs();
+        for (int i = 0; i < dirs.length; i++) {
+
+            // omit the mount point where CLASSIC_LOCATION lies (would be duplicate)
+            if (i == 0)
+                continue;
+
+            Location location = new Location();
+            location.id = i;
+            location.mediaPath = dirs[i];
+            location.cachePath = cacheDirs[i];
+            retList.add(location);
+        }
+
+        // add current selection at the end, even if the storage is currently unavailable
+        if (selectedDir != CLASSIC_LOCATION.id && selectedDir >= dirs.length) {
+            Location location = new Location();
+            location.id = selectedDir;
+            location.mediaPath = null;
+            location.cachePath = null;
+            retList.add(location);
+        }
+
+        for (Location loc: retList) {
+            loc.currentSelection = (loc.id == selectedDir);
+            fillLocationInfo(loc); // fill in size & description info
+        }
+
+        return retList;
+    }
+
+    /**
+     * Set the new storage directory.
+     *
+     * This will change the settings and move files from the old to the new location.
+     * Therefore, this method might take a while to finish.
+     *
+     * @param id the ID of the new storage location
+     */
+    public final void setStorageDir(int id) {
+
+        int oldID = SettingsManager.instance().getStorageDir();
+
+        if (oldID == id)
+            return;
+
+        Location newLocation = lookupStorageLocation(id);
+
+        File newMediaDir = newLocation.mediaPath;
+
+        if (newLocation.mediaPath == null || !EnvironmentCompat.getStorageState(newMediaDir).equals(Environment.MEDIA_MOUNTED)) {
+            Log.i(DEBUG_TAG, "Selected storage dir is unavailable! " + newMediaDir);
+            return;
+        }
+
+        Location oldLocation = lookupStorageLocation(oldID);
+        if (oldLocation != null) {
+            AccountManager manager = new AccountManager(getContext());
+
+            try {
+                // move cached files from old location to new location (might take a while)
+
+                for (Account account: manager.getAccountList()) {
+                    DataManager dataManager = new DataManager(account);
+                    File oldAccountDir = new File(dataManager.getAccountDir());
+
+                    if (oldAccountDir.isDirectory()) {
+                        FileUtils.copyDirectoryToDirectory(oldAccountDir, newMediaDir);
+                    }
+                }
+                notifyAndroidGalleryDirectoryChange(FileUtils.listFiles(newMediaDir, null, true));
+
+            } catch (IOException e) {
+                Log.e(DEBUG_TAG, "Could not move cache to new location", e);
+                return;
+            }
+
+            // remove everything in the old cache directories (thumbnails, etc).
+            clearCache();
```

Comment:
No. And we don't need to. All the "seafile.domain (user@domain.com)" directories are copied over to the new location. Therefore the SQL databases remain valid.

(Only less important cache files, like the .json files and thumbnails are lost during a storage location change. But they don't have any representation in the SQL DBs)